# Propositional Bounds

In this notebook, we will extend the representation of truth values to improve the expressiveness of logic. These extensions allow LNNs to handle a greater amount of ambiguity and uncertainty of beliefs otherwise present in logical reasoning systems. 

### Learning Outcomes Of This Tutorial

1. Bounds On Beliefs
    1. Extending Classical Truths
    2. Bounds
    3. Inference with Bounds

## Bounds On Beliefs

### Extending Classical Truths

In [discrete mathematics](https://en.wikipedia.org/wiki/Discrete_mathematics#Logic), logical systems reason using a finite set of truth values, namely <strong style=color:#0e6027>TRUE</strong> or <strong style=color:#ba4e00>FALSE</strong>. LNNs extend the expressiveness of the logic by allowing truth values to operate outside of the classical `{0, 1}` range as follows:

&emsp;i. &emsp;truths can be any intermediate truth value between `[0, 1]`, </br>
&emsp;ii.&emsp;truths can be bounded by a range of real-values.

These two extensions allow for handling (_i._) greater uncertainty and (_ii._) greater ambiguity of truth values. Uncertainty pertains to a truth value that is neither _True_ nor _False_, but could instead be any indefinite value between $[0, 1]$. It is achieved by incorporating the theory of [real-valued logics](https://en.wikipedia.org/wiki/Fuzzy_logic) into operator calculations, which will be expanded on in later notebooks. Ambiguity pertains to doubt in the semantics under the current context, thereby catering to more than one interpretation. This extends the representation of intermediate values to operate over a range, i.e., by keeping track of lower and upper bounds of real-values, represented as `[L, U]`.

In LNN we refer to the representation of truth values as bounds on beliefs, or bounds for short. Due to their flexibility, the semantics of bounds cannot be interpreted as probabilities, but should instead be perceived as _beliefs_. That is, the extent to which a contextual application of the knowledge believes a particular symbol to be [truthful](https://en.wikipedia.org/wiki/Truth).

Previously in the [Propositional Logic](./0.%20Propositional%20Logic.ipynb) notebook, we learned how to define propositions and their corresponding truths using `Facts`. In this notebook, we will expand on those ideas and demonstrate how to perform logical inference with bounds. 

Recall the LNN API to define facts:

<center>
    <figure>
    <img src="./img/1/messi.png" width="280" />
        <figcaption>Figure 1. Defining a proposition with a True truth value of assignment.</figcaption>
 </figure>
</center>

In [1]:
from lnn import Proposition, Fact

Messi = Proposition("Messi")
Messi.add_data(Fact.TRUE)
Messi.print()

OPEN Proposition: Messi                                     TRUE (1.0, 1.0)



### Bounds

Bounds are a pairwise representation of a truth range defined as a tuple of lower and upper real-values. At their extrema, they explictly encode classical truths, <strong style=color:#0e6027>TRUE</strong> or <strong style=color:#ba4e00>FALSE</strong>, or any real-value in between the $[0, 1]$ range, i.e., <strong style=color:#697077>UNKNOWNs</strong>.

<center>
    <figure>
    <img src="./img/1/bounds.png" width="280" />
        <figcaption>Figure 2. Proposition with a bound representation of a truth value</figcaption>
 </figure>
</center>

Below, we show the explicit representation of `Facts` using bounds with real-numbers.

| | "True" | "False" | "Unknown" |
| :-- | :--: | :--: | :--: |
| Fact | TRUE | FALSE | UNKNOWN |
| Bounds | `(1.0, 1.0)` | `(0.0, 0.0)` | `(0.0, 1.0)` | 

In the LNN, we use the all-caps notation of facts to indicate that the representation is a bounded representation, i.e., _True = 1._ and _TRUE = [1., 1.]_

We can also visualize the bounded representation of truths as follows:

<center>
    <figure>
    <img src="./img/1/messi_bounds.png" width="280" />
        <figcaption>Figure 2. Defining a <strong>Messi</strong> proposition and assigning it a truth value of <strong style=color:#0e6027>TRUE</strong> using bounds representation</figcaption>
 </figure>
</center>

There is a one-to-one mapping between classical facts and bounds, but a one-to-many mapping from non-classical facts to bounds. This relationship is expanded on in greater depth in [section F.2 of the LNN paper](https://arxiv.org/pdf/2006.13155.pdf). We demonstrate an example of a non-classical state below:

In [2]:
from lnn import Proposition, Fact

Rain = Proposition("Rain")
Rain.add_data((.7, 1.0))
Rain.print()

OPEN Proposition: Rain                               APPROX_TRUE (0.7, 1.0)



### Inference with Bounds

Recall in the [Propositional Logic](./0.%20Propositional%20Logic.ipynb) notebook, we discussed how to infer truth values of an operand/operator using omni-directional inference. Likewise, we will demonstrate how arbitrary bound values of an operand/operator are inferred using inference.

#### Upward Inference

In [3]:
from lnn import And, Fact

# Rules
EPL = Proposition("English Premier League")
BornInEngland = Proposition("Born in England")
AND = And(EPL, BornInEngland)

# Data
EPL.add_data((1.0, 1.0))
BornInEngland.add_data((0.1, .4))

# Reasoning
AND.upward()
AND.print()

OPEN And: (English Premier League ∧ Born in England)  APPROX_FALSE (0.1, 0.4)



The above example demonstrates that you can have an absolute belief that an individual plays in the English Premier league but a low confidence that the individual was born in England.

#### Downward Inference

In [4]:
from lnn import Implies

# Rules
BornInArgentina = Proposition("Born in Argentina")
SpeaksSpanish = Proposition("Speaks Spanish")
IMPLIES = Implies(BornInArgentina, SpeaksSpanish)

# Data
BornInArgentina.add_data((0.9, 1.0))
IMPLIES.add_data((0.95, 1.0))

# Reasoning
IMPLIES.downward()
SpeaksSpanish.print()

OPEN Proposition: Speaks Spanish                     APPROX_TRUE (0.85, 1.0)



The example above demonstrates that you can have a strong (but not absolute) belief that all _Argentinians speak Spanish_, i.e. between $0.95$ and $1.0$. This can also be interpreted as a confidence in the belief: "_We believe, with a confidence between 95-100%, that all Argentinians speak Spanish_". This 5% uncertainty in the belief still caters for the semantics of absolute certainty, i.e., there may be other rules or facts present in the system that could cause our belief in the implication to collapse on the upper bound and become perfectly true, $[1.0, 1.0]$.

A downward inference from the highly-confident antecedent (left-hand operand, `BornInArgentina`) onto the consequent (right-hand operand, `SpeaksSpanish`), results in a high confidence output belief, $[0.85, 1.0]$. The output truth values are calculated using an [implication in real-value logics](https://en.wikipedia.org/wiki/T-norm#Residua_of_prominent_left-continuous_t-norms), which we will expand on in future notebooks.

### In Summary

In this notebook we learned the following:
- The LNN extension of classical truths to include real-values in the `[0, 1]` range.
- The extension of individual truths to a range of truth semantics that are still consistent with the logic.
- Representing truth values using a bounds representation.
- Performing omni-directional inference with bounds.

Congrats on completing the Propositional Bounds example 🎉. 
Next, we will learn about activation functions and how to perform learning to correct the parameters of an LNN model.

[<img src="./img/next.png" width="160" alt="Next Tutorial: Propositional Learning" />](./2.%20Propositional%20Learning.ipynb)